rfc/010-distributed-coordinator.md

```diff
@@ -0,0 +1,249 @@
+# RFC 010: Distributed coordinator
+
+## Background
+
+The Contrast Coordinator is a stateful service with a backend storage that can't be shared.
+If the Coordinator pod restarts, it loses access to the secret seed and needs to be recovered manually.
+This leads to predictable outages on AKS, where the nodes are replaced periodically.
+
+## Requirements
+
+* High availability: node failure, pod migration, etc. must not impact the availability of the `set` or `verify` flow.
+* Automatic recovery: node failure, pod migration, etc. must not require a manual recovery step.
+  * A newly started Coordinator that has recovered peers must eventually recover automatically.
+* Consistency: the Coordinator state must be strongly consistent.
+* Data-owner security: the mesh CA key must only ever be exposed to Coordinator instances.
+
+## Design
+
+### Security considerations
+
+In order to allow reasoning about the chain of trust, we adhere to the following invariants:
+
+1. If a Coordinator has seed share owners $S_1,...,S_n$ in its active manifest, one of the following is true:
+   * It created its secret seed locally and only shared it in encrypted form with these seed share owners.
+   * It received its secret seed from an authenticated seed share owner $S_i$.
+   * It received its secret seed from an authenticated Coordinator with an active manifest `M` listing exactly $S_1,...,S_n$ as seed share owners.
+2. For the active mesh CA key at a Coordinator with active manifest `M`, one of the following is true:
+   * It was locally generated by the Coordinator when `M` was set.
+   * It was received from an authenticated Coordinator with active manifest `M`.
+3. A Coordinator only shares its secret seed and its mesh CA key with an authenticated Coordinator.
+
+Taken together, these invariants ensure
+
+* data owner security: a mesh CA key is only ever used with a single manifest and isn't known to non-Coordinators.
+* seed share owner security: the seed configured for an active manifest can be traced back to the seed share owners in that manifest.
+
+See the [section on manifest changes](#manifest-changes) below regarding authentication of Coordinators.
+
+### Persistency changes
+
+The Coordinator uses a generic key-value store interface to read from and write to the persistent store.
+In order to allow distributed read and write requests, we only need to swap the implementation for one that can be used concurrently.
+
+```golang
+type Store interface {
+  Get(key string) ([]byte, error)
+  Set(key string, value []byte) error
```

Comment:
If only `CompareAndSwap` is atomic, do we have a race risk with `Get`/`Set` here?

Reply:
These store primitives are used for different purposes:

* `Get`: snapshot of the current state of a piece of data without any freshness guarantees
* `Set`: store content-addressed data (policies, manifests)
* `CompareAndSwap`: atomically advance the latest transition (i.e. active manifest)

We implicitly expect the store to be race free for concurrent Get/Set calls (no partial or corrupt data).

All of this has been defined originally for recovery persistence: https://github.com/edgelesssys/contrast/blob/main/rfc/004-recovery.md#storage-interface.